### ISSUE TITLE
Feature Request: Persistent Undo/Redo Across Sessions

### ISSUE THREAD

**Issue by amqndin** [microsoft/vscode#224682]
Hey VSCode team,

I'd love to see a feature where VSCode saves a history of all the changes I make in a file, even after closing it, restarting VSCode, or even rebooting my computer. 

This would mean I could:

* **Undo changes from previous sessions:** I often realize I made a mistake after closing a file, and it would be awesome to be able to undo it the next time I open it.
* **Recover work after a crash:** If VSCode or my computer crashes, I wouldn't lose my undo history.
* **Have peace of mind:** I'd feel more confident knowing I can always go back and undo something, no matter what happens.

I know some other editors like neovim this feature, and I think it would be a great addition to VSCode. It would make my workflow smoother and less stressful. And I think a lot of people would agree with me on this adding this feature.

Thanks for considering my suggestion!

**Comment by connor4312** [comment]
You may want to check out the Timeline view which does this (bottom of the Explorer tab)

**Comment by bpasero** [comment]
https://github.com/microsoft/vscode/issues/17977
https://github.com/microsoft/vscode/issues/15135

**Comment by amqndin** [comment]
> You may want to check out the Timeline view which does this (bottom of the Explorer tab)

whaaat, thanks i will definetly look into that

**Comment by amqndin** [comment]
> See also #17977 and #15135.

oh, apologies

**Comment by amqndin** [comment]
yeah, timeline feature actually tracks all of the changes, but how can go back to previous state in timeline by a keyboard shortcut? that would be amaznig. thanks in advance

**Comment by khemarato** [comment]
Yes please! Would love edit history to persist after close / restart.

**Comment by Th0rGarden** [comment]
Is this still planned? It seems about time in 2025 we could keep our undo history of all things... this would save everyone from so many migraines and wasted time. Often have to restart vscode or the computer for many reasons, or there may be a crash, but the lack of persistent undo/redo history (more granular than the save-state based timeline) presents significant challenges to getting things done IRL.

**Comment by amqndin** [comment]
> cudatext editor have Persistent Undo/Redo, and have a setting to enable saving the Persistent Undo/Redo to a file near your edited file , so the Persistent Undo/Redo stack is never deleted and works even if you change the folder name or change its location, this is a great option, i no longer worry about crashes or restarts , just start cudatext again and you find your self like if no restart happened, a permanent continuous session. i left vscode just for this option, and cudatext is super fast and lightweight too. i hope to see this robust option in vscode someday

i actually left vscode for the same reason!
